replay/data_preparator.py

```diff
@@ -186,6 +190,209 @@ def _reindex(self, df: DataFrame, entity: str):
+            .select(self.item_col, (sf.col("item_idx") + prev_item_count).alias("item_idx"))
+        )
+        self.item_col_2_index_map = self.item_col_2_index_map.union(new_items_map)
+
+        new_users = (
+            df.join(self.user_col_2_index_map, on=self.user_col, how="left_anti")
+            .select(self.user_col).distinct()
+        )
+        prev_user_count = self.user_col_2_index_map.count()
+        new_users_map = (
+            IndexerEstimator.get_map(new_users, self.user_col, "user_idx")
+            .select(self.user_col, (sf.col("user_idx") + prev_user_count).alias("user_idx"))
+        )
+        self.user_col_2_index_map = self.user_col_2_index_map.union(new_users_map)
+
+    def _transform(self, df: DataFrame) -> DataFrame:
+
+        if self.update_map_on_transform:
+            self._update_maps(df)
+
+        if self.item_col in df.columns:
+            remaining_cols = df.drop(self.item_col).columns
+            df = df.join(self.item_col_2_index_map, on=self.item_col, how="left").select(
+                sf.col("item_idx").cast("int").alias("item_idx"),
+                *remaining_cols,
+            )
+        if self.user_col in df.columns:
+            remaining_cols = df.drop(self.user_col).columns
+            df = df.join(self.user_col_2_index_map, on=self.user_col, how="left").select(
+                sf.col("user_idx").cast("int").alias("user_idx"),
+                *remaining_cols,
+            )
+        return df
+
+    def inverse_transform(self, df: DataFrame) -> DataFrame:
+        """
+        Convert DataFrame to the initial indexes.
+
+        :param df: DataFrame with numerical ``user_idx/item_idx`` columns
+        :return: DataFrame with original user/item columns
+        """
+        if "item_idx" in df.columns:
+            remaining_cols = df.drop("item_idx").columns
+            df = df.join(self.item_col_2_index_map, on="item_idx", how="left").select(
+                sf.col(self.item_col).cast(self.item_type).alias(self.item_col),
+                *remaining_cols,
+            )
+        if "user_idx" in df.columns:
+            remaining_cols = df.drop("user_idx").columns
+            df = df.join(self.user_col_2_index_map, on="user_idx", how="left").select(
+                sf.col(self.user_col).cast(self.user_type).alias(self.user_col),
+                *remaining_cols,
+            )
+        return df
+
+
+class IndexerEstimator(Estimator):
+
+    def __init__(self, user_col="user_id", item_col="item_id"):
+        """
+        Provide column names for indexer to use
+        """
+        self.user_col = user_col
+        self.item_col = item_col
+        self.user_col_2_index_map = None
+        self.item_col_2_index_map = None
+
+    @staticmethod
+    def get_map(df: DataFrame, col_name: str, idx_col_name: str) -> DataFrame:
+        uid_rdd = (
+            df.select(col_name).distinct()
+            .rdd.map(lambda x: x[col_name])
```

Comment:
Is 'lambda' really necessary here? After the select op the df already should have only one column.

Reply:
Yes, otherwise, after spark.createDataFrame() is executed, the column will become a structure.